tests/gis_tests/test_geoip2.py

```diff
@@ -0,0 +1,125 @@
+
+        # Non-string query should raise TypeError
+        self.assertRaises(TypeError, cntry_g.country_code, 17)
+        self.assertRaises(TypeError, cntry_g.country_name, GeoIP2)
+
+    def test03_country(self):
+        "Testing GeoIP country querying methods."
+        g = GeoIP2(city='<foo>')
+
+        fqdn = 'www.google.com'
+        addr = '12.215.42.19'
+
+        for query in (fqdn, addr):
+            for func in (g.country_code, g.country_code_by_addr, g.country_code_by_name):
+                self.assertEqual('US', func(query), 'Failed for func %s and query %s' % (func, query))
+            for func in (g.country_name, g.country_name_by_addr, g.country_name_by_name):
+                self.assertEqual('United States', func(query), 'Failed for func %s and query %s' % (func, query))
+            self.assertEqual({'country_code': 'US', 'country_name': 'United States'},
+                             g.country(query))
+
+    @skipUnless(HAS_GEOS, "Geos is required")
+    def test04_city(self):
+        "Testing GeoIP city querying methods."
+        g = GeoIP2(country='<foo>')
+
+        addr = '128.249.1.1'
+        fqdn = 'tmc.edu'
+        for query in (fqdn, addr):
+            # Country queries should still work.
+            for func in (g.country_code, g.country_code_by_addr, g.country_code_by_name):
+                self.assertEqual('US', func(query))
+            for func in (g.country_name, g.country_name_by_addr, g.country_name_by_name):
+                self.assertEqual('United States', func(query))
+            self.assertEqual({'country_code': 'US', 'country_name': 'United States'},
+                             g.country(query))
+
+            # City information dictionary.
+            d = g.city(query)
+            self.assertEqual('US', d['country_code'])
+            self.assertEqual('Houston', d['city'])
+            self.assertEqual('TX', d['region'])
+
+            geom = g.geos(query)
+            self.assertIsInstance(geom, GEOSGeometry)
+            lon, lat = (-95.4010, 29.7079)
+            lat_lon = g.lat_lon(query)
+            lat_lon = (lat_lon[1], lat_lon[0])
+            for tup in (geom.tuple, g.coords(query), g.lon_lat(query), lat_lon):
+                self.assertAlmostEqual(lon, tup[0], 4)
+                self.assertAlmostEqual(lat, tup[1], 4)
+
+    def test05_unicode_response(self):
+        "Testing that GeoIP strings are properly encoded, see #16553."
+        g = GeoIP2()
+        d = g.city("duesseldorf.de")
+        self.assertEqual('Düsseldorf', d['city'])
+        d = g.country('200.26.205.1')
+        # Some databases have only unaccented countries
+        self.assertIn(d['country_name'], ('Curaçao', 'Curacao'))
+
+    def test06_ipv6_query(self):
+        "Testing that GeoIP can lookup ipv6 addresses."
+
+        g = GeoIP2()
+        d = g.city('2a03:2880:2110:3f01:face:b00c::')
+        self.assertNotEqual(d, None)
```

Comment:
is assertNotEqual to None really the best we can do? other tests seems more specific.